exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/FragmentsRunner.java

```diff
@@ -0,0 +1,439 @@
+                                   final CountDownLatch latch, final FragmentSubmitFailures fragmentSubmitFailures) {
+    @SuppressWarnings("resource")
+    final Controller controller = drillbitContext.getController();
+    final BitControl.InitializeFragments.Builder fb = BitControl.InitializeFragments.newBuilder();
+    for(final PlanFragment planFragment : fragments) {
+      fb.addFragment(planFragment);
+    }
+    final BitControl.InitializeFragments initFrags = fb.build();
+
+    logger.debug("Sending remote fragments to \nNode:\n{} \n\nData:\n{}", assignment, initFrags);
+    final FragmentSubmitListener listener =
+        new FragmentSubmitListener(assignment, initFrags, latch, fragmentSubmitFailures);
+    controller.getTunnel(assignment).sendFragments(listener, initFrags);
+  }
+
+  /**
+   * Add planFragment into either of local fragment list or remote fragment map based on assigned Drillbit Endpoint node
+   * and the local Drillbit Endpoint.
+   *
+   * @param planFragment plan fragment
+   * @param localEndPoint local endpoint
+   * @param localFragmentList local fragment list
+   * @param remoteFragmentMap remote fragment map
+   */
+  private void updateFragmentCollection(final PlanFragment planFragment, final CoordinationProtos.DrillbitEndpoint localEndPoint,
+                                        final List<PlanFragment> localFragmentList,
+                                        final Multimap<CoordinationProtos.DrillbitEndpoint, PlanFragment> remoteFragmentMap) {
+    final CoordinationProtos.DrillbitEndpoint assignedDrillbit = planFragment.getAssignment();
+
+    if (assignedDrillbit.equals(localEndPoint)) {
+      localFragmentList.add(planFragment);
+    } else {
+      remoteFragmentMap.put(assignedDrillbit, planFragment);
+    }
+  }
+
+  /**
+   * Send remote intermediate fragment to the assigned Drillbit node.
+   * Throw exception in case of failure to send the fragment.
+   *
+   * @param remoteFragmentMap - Map of Drillbit Endpoint to list of PlanFragment's
+   */
+  private void scheduleRemoteIntermediateFragments(final Multimap<CoordinationProtos.DrillbitEndpoint, PlanFragment> remoteFragmentMap) {
+
+    final int numIntFragments = remoteFragmentMap.keySet().size();
+    final ExtendedLatch endpointLatch = new ExtendedLatch(numIntFragments);
+    final FragmentSubmitFailures fragmentSubmitFailures = new FragmentSubmitFailures();
+
+    // send remote intermediate fragments
+    for (final CoordinationProtos.DrillbitEndpoint ep : remoteFragmentMap.keySet()) {
+      sendRemoteFragments(ep, remoteFragmentMap.get(ep), endpointLatch, fragmentSubmitFailures);
+    }
+
+    final long timeout = RPC_WAIT_IN_MSECS_PER_FRAGMENT * numIntFragments;
+    if (numIntFragments > 0 && !endpointLatch.awaitUninterruptibly(timeout)) {
+      long numberRemaining = endpointLatch.getCount();
+      throw UserException.connectionError()
```

Comment:
Here is a good example of an unchecked exception not handled by the outer try/catch block.